sms_sponsorship/controllers/main.py

```diff
@@ -143,14 +143,16 @@ def step2_confirm_sponsorship(self, sponsorship_id=None, **kwargs):
         """ SMS step2 controller. Returns the sponsorship registration form."""
         sponsorship = request.env['recurring.contract'].sudo().browse(
             sponsorship_id)
-        if sponsorship.sms_request_id.state == 'step2':
+        if sponsorship.sms_request_id.state == 'step2' or \
+                sponsorship.state in ['active', 'waiting']:
             # Sponsorship is already confirmed
             return self.sms_registration_confirmation(sponsorship.id, **kwargs)
-        return self.make_response(
-            'recurring.contract',
-            model_id=sponsorship and sponsorship.id,
-            **kwargs
-        )
+        if sponsorship.state == "draft":
+            return self.make_response(
+                'recurring.contract',
+                model_id=sponsorship and sponsorship.id,
+                **kwargs
+            )
```

Comment:
Oops, my second comment was not posted, sorry ! I proposed to add an else clause to raise an error in any other situation.
```suggestion
            )
        else:
            raise NotFound()
```